storage/localstore/migration_test.go

```
@@ -0,0 +1,381 @@
+	}
+
+	// start the fresh localstore with the sanctuary schema name
+	db, err := New(dir, baseKey, nil)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	err = db.Close()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	DbSchemaCurrent = "foo"
+
+	// start the existing localstore and expect the migration to run
+	db, err = New(dir, baseKey, nil)
+	if !strings.Contains(err.Error(), errMissingTargetSchema.Error()) {
+		t.Fatalf("expected errMissingTargetSchema but got %v", err)
+	}
+
+	if shouldNotRun {
+		t.Errorf("migration ran but shouldnt have")
+	}
+}
+
+// TestMigrateSanctuaryFixture migrates an actual Sanctuary localstore
+// to the most recent schema.
+func TestMigrateSanctuaryFixture(t *testing.T) {
+
+	tmpdir, err := ioutil.TempDir("", "localstore-test")
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer os.RemoveAll(tmpdir)
+
+	dir := path.Join(".", "testdata", "sanctuary")
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	files, err := ioutil.ReadDir(dir)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	for _, f := range files {
+		err = copyFileContents(path.Join(dir, f.Name()), path.Join(tmpdir, f.Name()))
+		if err != nil {
+			t.Fatal(err)
+		}
+	}
+
+	baseKey := make([]byte, 32)
+	if _, err := rand.Read(baseKey); err != nil {
+		t.Fatal(err)
+	}
+
+	// start localstore with the copied fixture
+	db, err := New(tmpdir, baseKey, &Options{Tags: chunk.NewTags()})
+	if err != nil {
+		t.Fatal(err)
+	}
+	schemaName, err := db.schemaName.Get()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if schemaName != DbSchemaCurrent {
+		t.Fatalf("schema name mismatch, want '%s' got '%s'", DbSchemaCurrent, schemaName)
+	}
+
+	err = db.Close()
+	if err != nil {
+		t.Fatal(err)
+	}
+}
+
+func copyFileContents(src, dst string) (err error) {
+	in, err := os.Open(src)
+	if err != nil {
+		return err
+	}
+	defer in.Close()
+	out, err := os.Create(dst)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		err = out.Close()
+	}()
+	if _, err = io.Copy(out, in); err != nil {
+		return err
+	}
+	err = out.Sync()
```

This error is never returned.